app/controllers/sites/new.js

```diff
@@ -58,52 +58,83 @@ export default class CreateSitesNewController extends Controller {
     };
   }
 
+  mapValidationDetailsToErrors(validationDetails) {
+    return validationDetails.reduce((accumulator, detail) => {
+      accumulator[detail.context.key] = detail.message;
+      return accumulator;
+    }, {});
+  }
+
+  async saveFormData() {
+    const { address, primaryContact, secondaryContact, site, adminUnit } =
+      this.model;
+    address.fullAddress = combineFullAddress(address);
+
+    await primaryContact.save();
+    await secondaryContact.save();
+    await address.save();
+
+    site.contacts = [primaryContact, secondaryContact];
+    site.address = address;
+    await site.save();
+
+    const nonPrimarySites = await adminUnit.sites;
+
+    if (this.isPrimarySite) {
+      const previousPrimarySite = await adminUnit.primarySite;
+
+      if (previousPrimarySite) {
+        nonPrimarySites.push(previousPrimarySite);
+      }
+
+      adminUnit.primarySite = site;
+    } else {
+      nonPrimarySites.push(site);
+    }
+
+    await adminUnit.save();
+  }
+
   saveTask = task(async (event) => {
```

Comment:
This is not a saveTask. This is a 'saveAndValidateTask'.
Refactor this into an synchronous (no async) action handler for the submit event which does the validation and start the save task if and only if the validation passes.
```
@action
handleSubmit(e) {
e.preventDefault();
... call validate etc ...
... IF validation OK THEN this.saveTask.perform
```
The body of the saveTask is identical to the current implementation of 'saveFormData'.